hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/handlers/ContractUpdateHandler.java

```diff
@@ -102,12 +105,112 @@ public void handle(@NonNull final HandleContext context) throws HandleException
-        validateSemantics(toBeUpdated, context, op);
+        validateSemantics(toBeUpdated, context, op, accountStore);
         final var changed = update(toBeUpdated, context, op);
 
         context.serviceApi(TokenServiceApi.class).updateContract(changed);
     }
 
+    private void validateSemantics(
+            Account contract,
+            HandleContext context,
+            ContractUpdateTransactionBody op,
+            ReadableAccountStore accountStore) {
+        validateTrue(contract != null, INVALID_CONTRACT_ID);
+
+        if (op.hasAdminKey() && processAdminKey(op)) {
+            throw new HandleException(INVALID_ADMIN_KEY);
+        }
+
+        if (op.hasExpirationTime()) {
+            try {
+                context.attributeValidator().validateExpiry(op.expirationTime().seconds());
+            } catch (HandleException e) {
+                validateFalse(contract.expiredAndPendingRemoval(), CONTRACT_EXPIRED_AND_PENDING_REMOVAL);
+                throw e;
+            }
+        }
+
+        validateFalse(!onlyAffectsExpiry(op) && !isMutable(contract), MODIFYING_IMMUTABLE_CONTRACT);
+        validateFalse(reducesExpiry(op, contract.expirationSecond()), EXPIRATION_REDUCTION_NOT_ALLOWED);
+
+        if (op.hasMaxAutomaticTokenAssociations()) {
+            final var ledgerConfig = context.configuration().getConfigData(LedgerConfig.class);
+            final var entitiesConfig = context.configuration().getConfigData(EntitiesConfig.class);
+            final var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+            final var contractsConfig = context.configuration().getConfigData(ContractsConfig.class);
+
+            final long newMax = op.maxAutomaticTokenAssociations();
+
+            validateFalse(
+                    newMax > ledgerConfig.maxAutoAssociations(),
+                    REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
+
+            validateFalse(newMax < contract.maxAutoAssociations(), EXISTING_AUTOMATIC_ASSOCIATIONS_EXCEED_GIVEN_LIMIT);
+            validateFalse(
+                    entitiesConfig.limitTokenAssociations() && newMax > tokensConfig.maxPerAccount(),
+                    REQUESTED_NUM_AUTOMATIC_ASSOCIATIONS_EXCEEDS_ASSOCIATION_LIMIT);
+
+            validateTrue(contractsConfig.allowAutoAssociations(), NOT_SUPPORTED);
+        }
+
+        // validate expiry metadata
+        final var currentMetadata =
+                new ExpiryMeta(contract.expirationSecond(), contract.autoRenewSeconds(), contract.autoRenewAccountId());
+        final var updateMeta = new ExpiryMeta(
+                op.hasExpirationTime() ? op.expirationTime().seconds() : NA,
+                op.hasAutoRenewPeriod() ? op.autoRenewPeriod().seconds() : NA,
+                null);
+        context.expiryValidator().resolveUpdateAttempt(currentMetadata, updateMeta);
+
+        context.serviceApi(TokenServiceApi.class)
+                .assertValidStakingElectionForUpdate(
```

Comment:
@tinker-michaelj regarding this comment - https://github.com/hashgraph/hedera-services/pull/9379#discussion_r1381722155 :
The `assertValidStakingElection` was calling `validateStakedIdForCreation`. I see we have `validateStakedIdForUpdate` and I think we should call that instead. What do you think?